File: examples/java/basic/src/test/java/com/rpl/agent/basic/MirrorAgentTest.java
package com.rpl.agent.basic;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import org.junit.Test;

/**
 * Test class for MirrorAgent demonstrating cross-module agent invocation testing.
 *
 * <p>This test demonstrates:
 *
 * <ul>
 *   <li>Deploying multiple modules to IPC
 *   <li>Testing cross-module agent invocation
 *   <li>Verifying mirror agent behavior
 * </ul>
 */
public class MirrorAgentTest {

  @Test
  public void testMirrorAgent() throws Exception {
    // Tests cross-module agent invocation via mirror agent
    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Deploy GreeterModule first
      MirrorAgent.GreeterModule greeterModule = new MirrorAgent.GreeterModule();
      ipc.launchModule(greeterModule, new LaunchConfig(1, 1));
      String greeterModuleName = greeterModule.getModuleName();

      // Deploy MirrorModule with reference to GreeterModule
      MirrorAgent.MirrorModule mirrorModule = new MirrorAgent.MirrorModule(greeterModuleName);
      ipc.launchModule(mirrorModule, new LaunchConfig(1, 1));

      // Get agent manager and client for MirrorModule
      String mirrorModuleName = mirrorModule.getModuleName();
      AgentManager manager = AgentManager.create(ipc, mirrorModuleName);
      AgentClient mirrorAgent = manager.getAgentClient("MirrorAgent");

      // Test cross-module invocation
      String result = (String) mirrorAgent.invoke("TestUser");
      assertNotNull("Mirror agent should return a result", result);
      assertEquals("Mirror says: Hello, TestUser!", result);
    }
  }
}

File: examples/clj/test/com/rpl/agent/basic/mirror_agent_test.clj
(ns com.rpl.agent.basic.mirror-agent-test
  (:require
   [clojure.test :refer [deftest testing is]]
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]
   [com.rpl.agent.basic.mirror-agent :refer [GreeterModule create-mirror-module]]))

(deftest mirror-agent-test
  ;; Test cross-module agent invocation using mirror agents
  (System/gc)
  (testing "MirrorAgent example"
    (with-open [ipc (rtest/create-ipc)]
      ;; Launch both modules
      (rtest/launch-module! ipc GreeterModule {:tasks 1 :threads 1})
      (let [greeter-module-name (rama/get-module-name GreeterModule)
            mirror-module (create-mirror-module greeter-module-name)]
        (rtest/launch-module! ipc mirror-module {:tasks 1 :threads 1})

        (let [manager (aor/agent-manager ipc (rama/get-module-name mirror-module))
              mirror-agent (aor/agent-client manager "MirrorAgent")]

          (testing "invokes Greeter via mirror and adds prefix"
            (is (= "Mirror says: Hello, Alice!"
                   (aor/agent-invoke mirror-agent "Alice"))))

          (testing "invokes Greeter via mirror for different input"
            (is (= "Mirror says: Hello, Bob!"
                   (aor/agent-invoke mirror-agent "Bob")))))))))